commons-core/src/test/scala/com/avsystem/commons/serialization/json/JsonStringInputOutputTest.scala

```diff
@@ -0,0 +1,206 @@
+    deserialized.binary should be(item.binary)
+    deserialized.list should be(item.list)
+    deserialized.set should be(item.set)
+    deserialized.obj should be(item.obj)
+    deserialized.map should be(item.map)
+  }
+
+  test("handle plain numbers in JSON as Int, Long and Double") {
+    val json = "123"
+    read[Int](json) should be(123)
+    read[Long](json) should be(123)
+    read[Double](json) should be(123)
+
+    val maxIntPlusOne: Long = Int.MaxValue.toLong + 1
+    val jsonLong = maxIntPlusOne.toString
+    intercept[ReadFailure](read[Int](jsonLong))
+    read[Long](jsonLong) should be(maxIntPlusOne)
+    read[Double](jsonLong) should be(maxIntPlusOne)
+
+    val jsonLongMax = Long.MaxValue.toString
+    intercept[ReadFailure](read[Int](jsonLong))
+    read[Long](jsonLongMax) should be(Long.MaxValue)
+    read[Double](jsonLongMax) should be(Long.MaxValue)
+
+    val jsonDouble = Double.MaxValue.toString
+    intercept[ReadFailure](read[Int](jsonDouble))
+    intercept[ReadFailure](read[Long](jsonDouble))
+    read[Double](jsonDouble) should be(Double.MaxValue)
+  }
+
+  test("work with skipping") {
+    case class TwoItems(i1: CompleteItem, i2: CompleteItem)
+    implicit val skippingCodec = new GenCodec[TwoItems] {
+      override def read(input: Input): TwoItems = {
+        val obj = input.readObject()
+        obj.nextField().skip()
+        val i2 = GenCodec.read[CompleteItem](obj.nextField())
+        TwoItems(null, i2)
+      }
+
+
+      override def write(output: Output, value: TwoItems): Unit = {
+        val obj = output.writeObject()
+        GenCodec.write[CompleteItem](obj.writeField("i1"), value.i1)
+        GenCodec.write[CompleteItem](obj.writeField("i2"), value.i2)
+        obj.finish()
+      }
+    }
+
+    val item = TwoItems(completeItem(), completeItem())
+    val serialized = write(item)
+    val deserialized = read[TwoItems](serialized)
+
+    deserialized.i1 should be(null)
+    deserialized.i2.unit should be(item.i2.unit)
+    deserialized.i2.string should be(item.i2.string)
+    deserialized.i2.char should be(item.i2.char)
+    deserialized.i2.boolean should be(item.i2.boolean)
+    deserialized.i2.byte should be(item.i2.byte)
+    deserialized.i2.short should be(item.i2.short)
+    deserialized.i2.int should be(item.i2.int)
+    deserialized.i2.long should be(item.i2.long)
+    deserialized.i2.float should be(item.i2.float)
+    deserialized.i2.double should be(item.i2.double)
+    deserialized.i2.binary should be(item.i2.binary)
+    deserialized.i2.list should be(item.i2.list)
+    deserialized.i2.set should be(item.i2.set)
+    deserialized.i2.obj should be(item.i2.obj)
+    deserialized.i2.map should be(item.i2.map)
+  }
+
+
+  test("serialize and deserialize huge case classes") {
+    def cc() = TestCC(Random.nextInt(), Random.nextLong(), Random.nextInt(), Random.nextBoolean(), Random.nextString(Random.nextInt(300)), List.fill(Random.nextInt(300))('a'))
```

Comment:
Randomized tests are not cool. You can't really rely on them to tell you that something was fixed after it was broken, so you can't do TDD-style bugfixing.
Property based tests also suffer from this although they tend to scan large set of possible values so it's less harmful.

Reply:
You do understand that this is a port of a test you most probably reviewed before? :smiling_imp:  I'd rather PBT but it did not work very well for us as you can see ;) I'd rather we detect more issues over a larger number of runs and then write specific test cases for them (as I did for scientific notation) than have a beautiful test suite which only makes me sure I can serialize 42 as a Double.